### ISSUE TITLE
Program does not write to RAM

### ISSUE THREAD

**Issue by KenH2000** [twelho/sram-tester#1]
I wanted to add an address line test and discovered the program does not actually write and read the data from the RAM chip.  It is only setting bits on the I/O lines, then reading them back without storing the data.  Every RAM chip tested is falsely identified as PASSING.
  
The following routine runs two loops, one to write the numbers to addresses and a second to read back the results from the tested addresses.  If the data were being stored on the RAM, the two loops should write and read back the same numbers stored at the tested addresses, but the loops do not agree.  The data written does not equal the data read.

````c
void addrtest(){

    uint32_t firstError = 0;
    uint32_t lastError = 0;
    uint64_t errorCount = 0;
    size_t pattern;
    uint32_t addr;
  // write to all address lines 

for (uint32_t i=0;i<NA; i++) {    
    addr = bit(i);
    pattern = (i+1);
    printBinary(pattern);
    Serial.print(" written at ");
    Serial.print(addr,HEX);
    Serial.println();      
    // Write test pattern to the SRAM
    writeData(addr, pattern);
    delay(100);
}

// read all address lines back
for (uint32_t i=0;i<NA; i++) {    
  addr = bit(i);
  pattern = (i+1);
  // Read data from the SRAM
  size_t data = readData(addr);
  delay(100);
      // Verify
      if (data != pattern) {
        lastError = addr;
        if (errorCount++ == 0) {
          firstError = addr;
        }
        
        Serial.print("Error at address 0x");
        Serial.print(addr, HEX);
        Serial.print(" - Got: ");
        printBinary(data);
        Serial.print(", expected: ");
        printBinary(pattern);
        Serial.println();
      }
}
}
````

**Comment by twelho** [comment]
This might be true, I did verify that the readings failed if the RAM chip is not present, but the chip might introduce enough capacitance such that the program is able to read its own measurement back. Your results would indicate that, assuming that `readData` is able to read data from the chip, writing the pattern does not work properly. I don't have the access to the chips nor the test setup anymore (the chips are in use), so could you find out what is wrong using your setup? Also, just as a sanity check, is the chip connected properly, including the write pin (and its inversion status)?

**Comment by KenH2000** [comment]
Unfortunately, the data returned by readData does not appear to be correct either.  I have an EPROM programmer that can read the RAM contents.  The RAM I'm using is NVRAM, so the contents stay the same when the chip is powered down.  The contents read by the EPROM programmer are always the same from chip to chip.  The data read on the Arduino does not match the contents read by the EPROM programmer.

I've tried a few things to troubleshoot the problem, but nothing worked.
1 Set up the OE as OUTPUT and changed its state from HIGH to LOW to HIGH prior to writing.
2 added different delays in parts of the write routine and read routine.
The wiring looks correct.  I don't have a scope or really know how to analyze the signals.  It is probably some kind of timing issue but I'm not sure what else to try.  It probably needs a capacitor at the chip between VCC and GND, but I don't know if that will solve the issue.
If I figure something else out I will post here.

**Comment by twelho** [comment]
I remember testing the application by intentionally causing errors by disconnecting data / address / write enable wires, and all the failures were correctly detected, indicating that the chip must have stored the values. Are you sure that your NVRAM chips work the same way as the legacy 2114 chips this is designed for? Based on your results, where both reading and writing fail, are you sure that the voltages are correct? The 2114 chips run at 5 V, which is why one can directly connect them to an Arduino. If your chips are 3.3 V, then they might write fine (despite the overvoltage), but the Arduino can't read them. Otherwise, if you're sure that the wiring is right, then I'd guess it is indeed a timing issue.

**Comment by KenH2000** [comment]
I've abandoned using this sketch as it neither reads or writes the data from or to the SRAM.  

I found a post for reading an EPROM that returns 100% accurate results when reading EPROMS.  
(https://danceswithferrets.org/geekblog/?p=315)  

I made some modifications and it is partially working to read and write to SRAMs.  
But it still needs some work -- as READing the levels back out is only partially successful.  

Since the EPROM reading program works so well at reading EPROMs, I'm convinced the problem with reading SRAMs is that the logic levels are NMOS or CMOS.  Those chips have different, ambiguous logic levels when compared to TTL, which is probably confusing the Arduino when reading the chip.
